[visionOS] Media: “…” options menu doesn’t present (when in full screen)

rr-codes opened the pull request:
#### 4776571d7bd69cdb89bb3315120497f74cc23064
<pre>
[visionOS] Media: “…” options menu doesn’t present (when in full screen)
<a href="https://bugs.webkit.org/show_bug.cgi?id=259589">https://bugs.webkit.org/show_bug.cgi?id=259589</a>
rdar://112867127

Reviewed by Tim Horton and Aditya Keerthi.

Instead of preventing any ornament from showing at all when in full screen, only hide/show the
ornaments that are present in the window when entering/exiting full screen.

* Source/WebKit/Platform/spi/visionos/MRUIKitSPI.h:
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(-[WKMRUIPlatterOrnamentProperties initWithDepthDisplacement:windowAlpha:]):
(-[WKFullScreenParentWindowState initWithWindow:]):
(-[WKFullScreenParentWindowState ornamentProperties]):
(-[WKFullScreenWindowController _setOrnamentsHidden:]):
(-[WKFullScreenWindowController _performSpatialFullScreenTransition:completionHandler:]):
(-[WKFullScreenParentWindowState ornamentDepths]): Deleted.

Canonical link: <a href="https://commits.webkit.org/266454@main">https://commits.webkit.org/266454@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6ee2d3c8c65574f2802b2cc2c65a548ff051e7e0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13850 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/14165 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/14500 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/15588 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/13158 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16673 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/14247 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/15833 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/14019 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14636 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11738 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/16293 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11923 "Passed tests") | | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/19537 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12999 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12662 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15877 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/13195 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/11072 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/12466 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3364 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16798 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13038 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1601,6 +1597,7 @@ - (void)_performSpatialFullScreenTransition:(BOOL)enter completionHandler:(Compl
             [ornament _setDepthDisplacement:originalDepth + kIncomingWindowZOffset];
 
         [UIView animateWithDuration:kWindowTranslationDuration delay:0 options:UIViewAnimationOptionCurveEaseInOut animations:^{
+            ornament.viewController.view.window.alpha = enter ? 0.0 : 1.0;
```
This should be done in the fade animation block.

pxlcoder replied:
Also the incoming/outgoing fade is different.

rr-codes replied:
fixed!

hortont424 commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1562,6 +1560,13 @@ - (void)_configureSpatialFullScreenTransition
+{
+    MRUIPlatterOrnamentManager *ornamentManager = window.windowScene._mrui_platterOrnamentManager;
+    for (MRUIPlatterOrnament *ornament in ornamentManager.ornaments)
+        ornament.viewController.view.window.alpha = visible ? 1.0 : 0.0;
```
Should we remember which ones we touched on the way in, and only un-touch those on the way out? What if the app added a hidden one in the meantime? (probably not a problem, "added a visible one" would be more likely but not a problem with your existing code).

rr-codes replied:
Ya, I suppose I can just do what we already do with ornamentDepths

pxlcoder replied:
I would reuse the `ornamentDepths` map. Also, passing in the window is unnecessary here, since both windows are in the same scene.

pxlcoder commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1562,6 +1560,13 @@ - (void)_configureSpatialFullScreenTransition
     });
 }
 
+- (void)_setOrnamentsVisabilityForWindow:(UIWindow *)window visible:(BOOL)visible
```
`Visability` -> `Visibility`

pxlcoder commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -467,6 +467,30 @@ - (BOOL)_shouldResizeWithScene
 
 @end
 
+@interface WKOrnamentProperties : NSObject
```
`WKMRUIPlatterOrnamentProperties`?

pxlcoder commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -500,20 +523,21 @@ - (id)initWithWindow:(UIWindow *)window
-    for (MRUIPlatterOrnament *ornament in ornamentManager.ornaments)
-        [_ornamentDepths setObject:@(ornament._depthDisplacement) forKey:ornament];
+    for (MRUIPlatterOrnament *ornament in ornamentManager.ornaments) {
+        WKOrnamentProperties *properties = [[WKOrnamentProperties alloc] initWithDepthDisplacement:ornament._depthDisplacement windowAlpha:ornament.viewController.view.window.alpha];
```
This is a memory leak.

pxlcoder commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1562,6 +1586,15 @@ - (void)_configureSpatialFullScreenTransition
     });
 }
 
+- (void)_setOrnamentsVisibility:(BOOL)visible
```
Should be `_setOrnamentsHidden` or `_setOrnamentsVisible`.

pxlcoder commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1562,6 +1586,15 @@ - (void)_configureSpatialFullScreenTransition
 
+- (void)_setOrnamentsVisibility:(BOOL)visible
+{
+    WKFullScreenParentWindowState *originalState = _parentWindowState.get();
```
This variable is unnecessary, you can just use square bracket syntax with `originalState`.

pxlcoder commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1562,6 +1586,14 @@ - (void)_configureSpatialFullScreenTransition
+{
+    for (MRUIPlatterOrnament *ornament in [_parentWindowState ornamentProperties]) {
+        CGFloat originalAlpha = [[[_parentWindowState ornamentProperties] objectForKey:ornament] windowAlpha];
+        ornament.viewController.view.window.alpha = hidden ? 0.0 : originalAlpha;
```
This could be rewritten to avoid map lookup when `hidden` is true, but not a very big deal.

pxlcoder commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -467,6 +467,30 @@ - (BOOL)_shouldResizeWithScene
+@interface WKMRUIPlatterOrnamentProperties : NSObject
+
+@property (nonatomic) CGFloat depthDisplacement;
+@property (nonatomic) CGFloat windowAlpha;
```
`@property (nonatomic, readonly)`

pxlcoder commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -500,20 +523,21 @@ - (id)initWithWindow:(UIWindow *)window
-    for (MRUIPlatterOrnament *ornament in ornamentManager.ornaments)
-        [_ornamentDepths setObject:@(ornament._depthDisplacement) forKey:ornament];
+    for (MRUIPlatterOrnament *ornament in ornamentManager.ornaments) {
+        auto properties = adoptNS([[WKMRUIPlatterOrnamentProperties alloc] initWithDepthDisplacement:ornament._depthDisplacement windowAlpha:ornament.viewController.view.window.alpha]);
```
Might be slightly cleaner if the initializer was `initWithOrnament:` and the properties were read in the initializer.

rr-codes committed:
[visionOS] Media: “…” options menu doesn’t present (when in full screen)
https://bugs.webkit.org/show_bug.cgi?id=259589
rdar://112867127

Reviewed by Tim Horton and Aditya Keerthi.

Instead of preventing any ornament from showing at all when in full screen, only hide/show the
ornaments that are present in the window when entering/exiting full screen.

* Source/WebKit/Platform/spi/visionos/MRUIKitSPI.h:
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(-[WKMRUIPlatterOrnamentProperties initWithDepthDisplacement:windowAlpha:]):
(-[WKFullScreenParentWindowState initWithWindow:]):
(-[WKFullScreenParentWindowState ornamentProperties]):
(-[WKFullScreenWindowController _setOrnamentsHidden:]):
(-[WKFullScreenWindowController _performSpatialFullScreenTransition:completionHandler:]):
(-[WKFullScreenParentWindowState ornamentDepths]): Deleted.

Canonical link: https://commits.webkit.org/266454@main
